src/Microsoft.Azure.Batch.SoftwareEntitlement/NodeEntitlementReader.cs

```diff
@@ -0,0 +1,162 @@
+                ValidateLifetime = true,
+                RequireExpirationTime = true,
+                RequireSignedTokens = _signingKey != null,
+                ClockSkew = TimeSpan.FromSeconds(60),
+                IssuerSigningKey = _signingKey,
+                ValidateIssuerSigningKey = true,
+                TokenDecryptionKey = _encryptionKey
+            };
+
+            try
+            {
+                var handler = new JwtSecurityTokenHandler();
+                var principal = handler.ValidateToken(tokenString, validationParameters, out var token);
+
+                // Set standard claims from token: NotBefore, NotAfter and Issuer
+                var result = new NodeEntitlements()
+                    .FromInstant(new DateTimeOffset(token.ValidFrom))
+                    .UntilInstant(new DateTimeOffset(token.ValidTo))
+                    .WithIssuer(token.Issuer);
+
+                // We don't expect multiple audiences to appear in the token
+                var audience = (token as JwtSecurityToken)?.Audiences?.SingleOrDefault();
+                if (audience != null)
+                {
+                    result = result.WithAudience(audience);
+                }
+
+                var virtualMachineIdClaim = principal.FindFirst(Claims.VirtualMachineId);
+                if (virtualMachineIdClaim != null)
+                {
+                    result = result.WithVirtualMachineId(virtualMachineIdClaim.Value);
+                }
+
+                var cpuCoreCountClaim = principal.FindFirst(Claims.CpuCoreCount);
+                if (cpuCoreCountClaim != null && int.TryParse(cpuCoreCountClaim.Value, out int maxCpuCoreCount))
+                {
+                    result = result.WithCpuCoreCount(maxCpuCoreCount);
+                }
+
+                foreach (var applicationClaim in principal.FindAll(Claims.Application))
+                {
+                    result = result.AddApplication(applicationClaim.Value);
+                }
+
+                foreach (var ipClaim in principal.FindAll(Claims.IpAddress))
+                {
+                    if (IPAddress.TryParse(ipClaim.Value, out var parsedAddress))
+                    {
+                        result = result.AddIpAddress(parsedAddress);
+                    }
+                    else
+                    {
+                        return InvalidTokenError($"Invalid IP claim: {ipClaim.Value}");
+                    }
+                }
+
+                var entitlementIdClaim = principal.FindFirst(Claims.EntitlementId);
+                if (entitlementIdClaim != null)
+                {
+                    result = result.WithIdentifier(entitlementIdClaim.Value);
+                }
+
+                var batchAccountIdClaim = principal.FindFirst(Claims.BatchAccountId);
+                if (batchAccountIdClaim != null)
+                {
+                    result = result.WithBatchAccountId(batchAccountIdClaim.Value);
+                }
```

Comment:
This pattern is repetitive and the method is getting a bit too long - I wonder if we can simplify by introducing a helper method ...

Reply:
That's something I tried but didn't come out as clean as I'd like. Got a couple of ideas, but I'd prefer to do these as a separate PR to make it easier to see if it's worth it.

Reply:
What about a local function that would have access to read `principal` and modify `result`):

``` csharp
void ReadClaim(string claimId, Func<NodeEntitlements, string, NodeEntitlements) action)
{
    var claim = principal.FindFirst(claimId);
    if (claim != null)
    { 
        result = action(result, claim.Value);
    }
}

ReadClaim(Claims.BatchAccountId, (e,s) => e.WithBatchAccountId(s));
ReadClaim(Claims.PoolId, (e,s) => e.WithPoolId(s));
ReadClaim(Claims.JobId, (e,s) => e.WithJobId(s));
```